lib/apiservers/service/restapi/handlers/common.go

```diff
@@ -40,71 +41,82 @@ type buildDataParams struct {
 	thumbprint      *string
 	datacenter      *string
 	computeResource *string
+	vchID           *string
 }
 
-func buildData(op trace.Operation, params buildDataParams, principal interface{}) (*data.Data, error) {
-	d := data.Data{
+func buildDataAndValidateTarget(op trace.Operation, params buildDataParams, principal interface{}) (*data.Data, *validate.Validator, error) {
+	data := &data.Data{
 		Target: &common.Target{
 			URL: &url.URL{Host: params.target},
 		},
 	}
 
 	if c, ok := principal.(Credentials); ok {
-		d.Target.User = c.user
-		d.Target.Password = &c.pass
+		data.Target.User = c.user
+		data.Target.Password = &c.pass
 	} else {
-		d.Target.CloneTicket = principal.(Session).ticket
+		data.Target.CloneTicket = principal.(Session).ticket
+	}
+
+	if err := data.HasCredentials(op); err != nil {
+		return data, nil, util.NewError(http.StatusUnauthorized, "Invalid Credentials: %s", err)
 	}
 
 	if params.thumbprint != nil {
-		d.Thumbprint = *params.thumbprint
+		data.Thumbprint = *params.thumbprint
+	}
+
+	if params.computeResource != nil {
+		data.ComputeResourcePath = *params.computeResource
 	}
 
+	if params.vchID != nil {
+		data.ID = *params.vchID
+	}
+
+	// TODO (#6032): clean this up
+	var validator *validate.Validator
 	if params.datacenter != nil {
-		validator, err := validateTarget(op, &d)
+		v, err := validate.NewValidator(op, data)
 		if err != nil {
-			return nil, util.WrapError(http.StatusInternalServerError, err)
+			return data, nil, util.NewError(http.StatusBadRequest, "Validation Error: %s", err)
 		}
 
 		datacenterManagedObjectReference := types.ManagedObjectReference{Type: "Datacenter", Value: *params.datacenter}
 
-		datacenterObject, err := validator.Session.Finder.ObjectReference(op, datacenterManagedObjectReference)
+		datacenterObject, err := v.Session.Finder.ObjectReference(op, datacenterManagedObjectReference)
 		if err != nil {
-			return nil, util.WrapError(http.StatusNotFound, err)
+			return nil, nil, util.WrapError(http.StatusNotFound, err)
 		}
 
-		d.Target.URL.Path = datacenterObject.(*object.Datacenter).InventoryPath
-	}
+		data.Target.URL.Path = datacenterObject.(*object.Datacenter).InventoryPath
 
-	if params.computeResource != nil {
-		d.ComputeResourcePath = *params.computeResource
-	}
+		// Copied from NewValidator
```

Comment:
I really don't like this - I can see you're trying to limit the changes to not touch anything outside of the server code, but it's really fragile.

I think our neater options are:
1. Update NewValidator to support a datacenter moref 
2. Add a utility method to construct a session from `data.Data`, resolve the datacenter with that session, then use it to construct a Target url with /datacenter
3. Proper refactoring of validator 

However... if confining the changes to this package then I think this needs the following:
```
dc, ok := datacenterObject.(*object.Datacenter)
if !ok {
   return data, nil, util.NewError(http.StatusBadRequest, "Validation Error: datacenter parameter is not a datacenter moref")

v.Session.Finder.SetDatacenter(dc)
v.DatacenterPath = dc.Name()
```

The `finder.SetDatacenter` operation is what sets the scoping for all subsequent moref/name resolution queries.

Reply:
> I really don't like this - I can see you're trying to limit the changes to not touch anything outside of the server code, but it's really fragile.

This is definitely something I'm torn about.

However, I think confining the changes to server code for 1.3 greatly reduces the overall risk of the change. As soon as `master` is open for post-1.3 changes, I'd like to explore option 1 or 3. (I'm not sure 2 works without 3, since to achieve 2 we'd need to refactor validator enough that it can accept an existing session.)

I'm also not sure any of these options remove the ugly aspects of this change. For example, achieving option 1 still requires creating a session, using it to resolve a datacenter moref into a path, and updating the session to use that datacenter path. That is, the fragile code for updating a session which has already been connected would still exist ― just in a different place. (I think that different place is a _better_ place in the long run, but I don't think it's worth the risk for 1.3.)

> However... if confining the changes to this package then I think this needs the following:

Is this code something you wrote for this review, or was it borrowed from somewhere?

> `dc, ok := datacenterObject.(*object.Datacenter)`

I think it would be good to be defensive like this, but it does not seem like `!ok` could ever happy with the way `object.NewReference` is currently implemented. Is that correct?

(Happy to add this; just trying to make sure I haven't misunderstood something.)

> `v.Session.Finder.SetDatacenter(dc)`

Good catch!

> `v.DatacenterPath = dc.Name()`

This seems slightly different from `strings.TrimPrefix(data.Target.URL.Path, "/")`, which is what's used in `validator`:

https://github.com/vmware/vic/blob/d860ced9c3e1362ef04447798a28025cef29c94a/lib/install/validate/validator.go#L136

Might there be cases where `.Name()` doesn't produce the right value here?

Reply:
a. You're correct that the cast will always succeed as written, but that's dependent on code from a different repo so I tend towards defensive.

b. It is slightly different because the source is coming from a different location - it may be called a path but if you look at `validator.go:164` we actually ensure there's no path separators because it was a common error to add cluster, etc, as well. Finder will cope with a path, and may in fact do better with it, but I don't know if anything else will break if we allow a DatacenterPath with multiple segments.